### PR TITLE
Added new method RegisterEditorResources to register widget editing resources instead of calling BuildEditorAsync

### LINKED ISSUE #3250: Widget Display Driver being called multiple times in Admin
Using `1.0.0-beta3-70622`

In our Orchard project, we have a fairly large amount of Widgets, most of which have to make external API calls to retrieve the data they need. Both when they are in their edit state as well as they're design state.

When creating or editing a page, Orchard calls `EditAsync` on every DisplayDriver on every Widget that is currently active in the system. For us, this means all the external API calls happen if a page is created or edited (regardless if they are used on the page).

I assumed Orchard was caching these for when the user added the widget to the page, however, the `EditAsync` method is called again when that happens. 

I realise our situation is unique, and Orchard was designed to have these methods read data from its own database, however it would be fantastic if this can be mitigated somehow. Either by removing the initial call to `EditAsync` as it does not seem like it is required? Or when that initial call happens, can a flag be present which can be checked, so we don't make external API calls unless required.

### LINKED ISSUE #4223: Widget scripts being added to the dom before widget is added.
I am not too sure how to report this issue but here is my test case:

1- Run blog recipe
2- Add ContentPickerField to Paragraph and Page Content Types
3- Create a new Page ContentItem and the init script for the ParagraphContentPickerField get's rendered before adding a widget (in this case the script was throwing an error)

image

I have fixed the VueMultiselectWrapper script so it does not throw an error if the element is null but I feel like those init scripts should not be added to the dom unless the user adds the widget to the type.

### LINKED ISSUE #5149: In admin theme, changing Location of buttons using placement.json shows lots of buttons in UI
In Admin theme, 

When "Actions" zone is added in `Layout.cshtml `  as following

 ```
<div class="ta-content">
            @await RenderSectionAsync("Messages", required: false)
            @await RenderSectionAsync("Breadcrumbs", required: false)

            @await RenderBodyAsync()

            @await RenderSectionAsync("Actions", required: false)


        </div>
```

And use `placement.json` as below

**placement.json**
```
{
  "Content_PublishButton": [
    {
      "place": "/Actions:5"
    }
  ],
  "Content_SaveDraftButton": [
    {
      "place": "/Actions:6"
    }
  ],
  "ContentPreview_Button": [
    {
      "place": "/Actions:6"
    }
  ]
}
```

Edit Content shows lots of buttons in UI.  Publish, Save Draft and Preview button would be seen only once. 

image

### PR DESCRIPTION
**Pull request by ns8482e**
Continuation with PR #5499 

Fixes #5149 
Fixes #3250 
Fixes #4223

If content item contains parts like `FlowPart`, `BagPart`, `WidgetListPart` mock editor is created by calling `BuildEditorAsync` for all the widgets to register resources (scripts and styles required for widget editing) with the Page.

`BuildEditorAsync `Invokes all the Display Handlers registered with Service collection. 
`ContentPreviewDriver `and `ContentsDriver `are such display handlers that invoked during `BuildEditorAsync`, that are responsible for creating action buttons like `Preview`, `Save Draft`

These mock editor for widgets are not intended to render But when `placement.json` or `<zone>` tag helper is used to place shapes on specific zone, shapes from mock editor are also placed on these zones and rendered.

If mock editor for widget contains content specific initialization `<script>` which are not intended to be included as resources for widget editing but are included in the Page unnecessarily  and in some cases throws java script error or adhoc fix like below is required. https://github.com/OrchardCMS/OrchardCore/blob/a5e11d392ac191b6b9d1baa99a1753fb1fdbf4fa/src/OrchardCore.Modules/OrchardCore.Liquid/Views/LiquidPart.Edit.cshtml#L20-L24

Also it increases the response size.

The solution is, 

Added new Method `Task RegisterEditorResources(ContentItem content);` 
Added property `public bool IsMockRendering ` on RazorPage

New method, `RegisterEditorResources `is similar to `BuildEditor`, except it renders everything using `MockLayout` instead of `Layout` from LayoutAccessor, so that any mock UI rendered are not placed on Page `Layout` zones 

Widget can now use `if(!IsMockRendering)` to include `DOM` `<script>`, only if it's not mock rendering 
Also added `IsMockRendering` condition for `BagPart.Edit.cshtml`, `FlowPart.Edit.cshtml`, `WidgetsListPart.Edit.cshtml`, that reduced the file size to `376 KB` from `505 KB` in my case.

image

### CODE REVIEW

**Comment by sebastienros** [comment]
You can't assume the edit shape will always be with `_Edit` and there could even be multiple shapes. 

What about being explicit about being able to render resources, either with a method that would take an IResourceManager and be called for widget drivers, or even as a different service that widgets can also implement. Or a custom display type so we can create a template that will only render the resources? I kind of like the display type approach, as we could just refactor existing widgets by extracting the resources section to a different template. And invoke this display type in the current code where we try to extract them. Or support both models, driver based, and template based.

**Comment by ns8482e** [comment]
> You can't assume the edit shape will always be with `_Edit`

I just did line-to-line copy from `BuildEditorAsync`. New method, `RegisterEditorResources` is same as `BuildEditorAsync`, except it renders everything using `MockLayout`, so nothing gets written to page output.
https://github.com/OrchardCMS/OrchardCore/blob/d599a9a654849d98bb9374a96d5db93fc0e32ef5/src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplayManager.cs#L117



> custom display type so we can create a template that will only render the resources? I kind of like the display type approach, as we could just refactor existing widgets

Then, the template  definition is required to for every `ContentParts` and `ContentFields` (out-of-box and any custom developed)  that makes a content type of the widget in Orchard Core.

This PR  is on similar approach but using  `Editor` shapes instead of requiring to define new display type.

> we could just refactor existing widgets by extracting the resources section to a different template.

This PR provides `IsMockRendering` flag within `Editor` shape to differentiate in case if refactoring required.

**Comment by ns8482e** [comment]
> ~This PR provides `IsMockRendering` flag within `Editor` shape to differentiate in case if refactoring required to exclude `script `or `style `~

Removing `IsMockRendering` property from RazorPage as it seems unnecessary. `<html>` content is already excluded  the issue is only with inline `<script>` tag. 

I think adding `ViewContext` property with `[ViewContext]` attribute on `ScriptTagHelper` would allows access to underlying shape and condition be checked  for `Mocklayout`, if the execution is for `RegisterEditorResources` or not

**Comment by agriffard** [comment]
@sebastienros Can you please review this PR?

File: src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplayManager.cs [hunk 1]
```diff
@@ -176,5 +180,49 @@ public async Task<IShape> UpdateEditorAsync(ContentItem contentItem, IUpdateMode
 
             return context.Shape;
         }
+
+        //Build a Mock editor Shape to register styles and script to document
+        //Shape result does not render.
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplayManager.cs]: Add a space after the `//`

File: src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplayManager.cs [hunk 2]
```diff
@@ -176,5 +180,49 @@ public async Task<IShape> UpdateEditorAsync(ContentItem contentItem, IUpdateMode
...
+
+            var stereotype = contentTypeDefinition.GetSettings<ContentTypeSettings>().Stereotype;
+
+            var actualShapeType = (stereotype ?? "Content") + "_Edit";
+
+            dynamic itemShape = await CreateContentShapeAsync(actualShapeType);
+            itemShape.ContentItem = contentItem;
+
+            // adding an alternate for [Stereotype]_Edit__[ContentType] e.g. Content-Menu.Edit
+            ((IShape)itemShape).Metadata.Alternates.Add(actualShapeType + "__" + contentItem.ContentType);
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplayManager.cs]: Star comment with an uppercase

File: src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplayManager.cs [hunk 3]
```diff
@@ -176,5 +180,49 @@ public async Task<IShape> UpdateEditorAsync(ContentItem contentItem, IUpdateMode
...
+
+            //mock layout
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplayManager.cs]: Space + uppercase

File: src/OrchardCore/OrchardCore.ResourceManagement/TagHelpers/ScriptTagHelper.cs [hunk 4]
```diff
@@ -261,6 +267,21 @@ public override async Task ProcessAsync(TagHelperContext context, TagHelperOutpu
...
             {
                 // Custom script content
+                var shapePrefix = string.Empty;
+                try
+                {
+                    shapePrefix = model?.Metadata?.Prefix as string;
+                }
+                catch (RuntimeBinderException)
+                {
+                    //Model is not a shape.
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.ResourceManagement/TagHelpers/ScriptTagHelper.cs]: Add a space

File: src/OrchardCore/OrchardCore.ResourceManagement/TagHelpers/ScriptTagHelper.cs [hunk 5]
```diff
@@ -261,6 +267,21 @@ public override async Task ProcessAsync(TagHelperContext context, TagHelperOutpu
...
+                }
+
+                if (!string.IsNullOrWhiteSpace(shapePrefix) && shapePrefix.StartsWith("MockLayout_"))
+                {
+                    //if mock redering do not render inline dom scripts
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.ResourceManagement/TagHelpers/ScriptTagHelper.cs]: Add a space + uppercase

File: src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplayManager.cs [hunk 6]
```diff
@@ -176,5 +180,49 @@ public async Task<IShape> UpdateEditorAsync(ContentItem contentItem, IUpdateMode
 
             return context.Shape;
         }
+
+        //Build a Mock editor Shape to register styles and script to document
+        //Shape result does not render.
+        public async Task RegisterEditorResources(ContentItem contentItem)
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.ContentManagement.Display/ContentDisplayManager.cs]: Normally async method names end with `Async`

**Pull request closed without merging**